pydantic/type_adapter.py

```diff
@@ -233,100 +208,130 @@ def __init__(
-        assert isinstance(self._validator, (SchemaValidator, PluggableSchemaValidator))
-        return self._validator
-
-    @cached_property
-    @_frame_depth(2)  # +2 for @cached_property + serializer(self)
-    def serializer(self) -> SchemaSerializer:
-        """The pydantic-core SchemaSerializer used to dump instances of the model."""
-        if not isinstance(self._serializer, SchemaSerializer):
-            self._init_core_attrs(rebuild_mocks=True)  # Do not expose MockValSer from public function
-        assert isinstance(self._serializer, SchemaSerializer)
-        return self._serializer
+            self.serializer = SchemaSerializer(self.core_schema, core_config)
 
+        # TODO: I think we should move this to the rebuild pattern?
+        if isinstance(self.core_schema, _mock_val_ser.MockCoreSchema):
+            self.core_schema = self.core_schema.rebuild()  # type: ignore[assignment]
+            self._init_core_attrs(ns_resolver=ns_resolver, force=True)
+
+        self._pydantic_complete = True
+        return True
+
+    @property
     def _defer_build(self) -> bool:
-        config = self._config if self._config is not None else self._model_config()
+        config = self._config if self._config is not None else self._model_config
         if config:
             return config.get('defer_build') is True
         return False
 
+    @property
     def _model_config(self) -> ConfigDict | None:
         type_: Any = _typing_extra.annotated_type(self._type) or self._type  # Eg FastAPI heavily uses Annotated
         if _utils.lenient_issubclass(type_, BaseModel):
             return type_.model_config
         return getattr(type_, '__pydantic_config__', None)
 
-    @_frame_depth(1)
+    def rebuild(
+        self,
+        *,
+        force: bool = False,
+        _parent_namespace_depth: int = 2,
+        _types_namespace: dict[str, Any] | None = None,
+    ) -> bool | None:
+        """Try to rebuild the pydantic-core schema for the adapter's type.
+
+        This may be necessary when one of the annotations is a ForwardRef which could not be resolved during
+        the initial attempt to build the schema, and automatic rebuilding fails.
+
+        Args:
+            force: Whether to force the rebuilding of the type adapter's schema, defaults to `False`.
+            _parent_namespace_depth: The depth level of the parent namespace, defaults to 2.
+            _types_namespace: The types namespace, defaults to `None`.
+
+        Returns:
+            Returns `None` if the schema is already "complete" and rebuilding was not required.
+            If rebuilding _was_ required, returns `True` if rebuilding was successful, otherwise `False`.
+        """
+        if not force and self._pydantic_complete:
+            return None
+        else:
+            if '_core_schema' in self.__dict__:
+                delattr(self, '_core_schema')  # delete cached value to ensure full rebuild happens
```

Comment:
Hmm is it correct as these were renamed?:
```suggestion
            if 'core_schema' in self.__dict__:
                delattr(self, 'core_schema')  # delete cached value to ensure full rebuild happens
```

Reply:
Good catch!